File: clox-pp/lexer.cpp
#include "lexer.h"

#include <cctype>
#include <cstring>

#include "token.h"
Lexer::Lexer(const std::string& source) : source{source}, start{&source[0]}, current{&source[0]}, line{1} {
}

Token Lexer::scanToken() {
    skipWhitespace();
    start = current;
    if (isAtEnd()) {
        return makeToken(TokenType::Eof);
    }
    char c = advance();
    switch (c) {
        case '(':
            return makeToken(TokenType::Left_paren);
        case ')':
            return makeToken(TokenType::Right_paren);
        case '{':
            return makeToken(TokenType::Left_brace);
        case '}':
            return makeToken(TokenType::Right_brace);
        case ';':
            return makeToken(TokenType::Semicolon);
        case ',':
            return makeToken(TokenType::Comma);
        case '.':
            return makeToken(TokenType::Dot);
        case '-':
            return makeToken(TokenType::Minus);
        case '+':
            return makeToken(TokenType::Plus);
        case '/':
            //            if (peekNext() == '/') {
            //                while (peek()!='\n' && !isAtEnd()) {advance();}
            //            }
            return makeToken(TokenType::Slash);
        case '*':
            return makeToken(TokenType::Star);
        case '!':
            return makeToken(match('=') ? TokenType::Bang_equal : TokenType::Bang);
        case '=':
            return makeToken(match('=') ? TokenType::Equal_equal : TokenType::Equal);
        case '<':
            return makeToken(match('=') ? TokenType::Less_equal : TokenType::Less);
        case '>':
            return makeToken(match('=') ? TokenType::Greater_equal : TokenType::Greater);
        case '"':
            while (peek() != '"' && !isAtEnd()) {
                if (peek() == '\n') {
                    ++line;
                }
                advance();
            }
            if (isAtEnd()) {
                return errorToken("Unterminated string.");
            }
            advance();
            return makeToken(TokenType::String);
        case '0':
        case '1':
        case '2':
        case '3':
        case '4':
        case '5':
        case '6':
        case '7':
        case '8':
        case '9':
            while (std::isdigit(peek())) {
                advance();
            }
            if (peek() == '.' && std::isdigit(peekNext())) {
                advance();
                while (std::isdigit(peek())) {
                    advance();
                }
            }
            return makeToken(TokenType::Number);

        case 'A':
        case 'B':
        case 'C':
        case 'D':
        case 'E':
        case 'F':
        case 'G':
        case 'H':
        case 'I':
        case 'J':
        case 'K':
        case 'L':
        case 'M':
        case 'N':
        case 'O':
        case 'P':
        case 'Q':
        case 'R':
        case 'S':
        case 'T':
        case 'U':
        case 'V':
        case 'W':
        case 'X':
        case 'Y':
        case 'Z':
        case 'a':
        case 'b':
        case 'c':
        case 'd':
        case 'e':
        case 'f':
        case 'g':
        case 'h':
        case 'i':
        case 'j':
        case 'k':
        case 'l':
        case 'm':
        case 'n':
        case 'o':
        case 'p':
        case 'q':
        case 'r':
        case 's':
        case 't':
        case 'u':
        case 'v':
        case 'w':
        case 'x':
        case 'y':
        case 'z':
        case '_':
            while (std::isalpha(peek()) || std::isdigit(peek()) || peek() == '_') {
                advance();
            }
            return makeToken(identifierType());
        default:
            return errorToken("Unexpected character.");
    }
}

const char* Lexer::getStart() const {
    return start;
}

const char* Lexer::getCurrent() const {
    return current;
}

int Lexer::getLine() const {
    return line;
}

bool Lexer::isAtEnd() const {
    return *current == '\0';
}

char Lexer::advance() {
    return *current++;
}

bool Lexer::match(char c) {
    if (isAtEnd()) {
        return false;
    }
    if (*current != c) {
        return false;
    }
    current++;
    return true;
}

void Lexer::skipWhitespace() {
    while (true) {
        char c = peek();
        switch (c) {
            case ' ':
            case '\r':
            case '\t':
                advance();
                break;

            case '\n':
                ++line;
                advance();
                break;

            case '/':
                if (peekNext() == '/') {
                    while (peek() != '\n' && !isAtEnd()) {
                        advance();
                    }
                } else {
                    return;
                }
                break;

            default:
                return;
        }
    }
}

char Lexer::peek() const {
    return *current;
}

char Lexer::peekNext() const {
    if (isAtEnd()) {
        return '\0';
    }
    return current[1];
}

Token Lexer::makeToken(TokenType type) const {
    return Token(type, *this);
}

Token Lexer::errorToken(const char* errorMsg) const {
    return Token(errorMsg, *this);
}

TokenType Lexer::identifierType() const {
    switch (*start) {
        case 'a':
            return checkKeyword(1, 2, "nd", TokenType::And);
        case 'c':
            return checkKeyword(1, 4, "lass", TokenType::Class);
        case 'e':
            return checkKeyword(1, 3, "lse", TokenType::Else);
        case 'f':
            if (current - start > 1) {
                switch (start[1]) {
                    case 'a':
                        return checkKeyword(2, 3, "lse", TokenType::False);
                    case 'o':
                        return checkKeyword(2, 1, "r", TokenType::For);
                    case 'u':
                        return checkKeyword(2, 1, "n", TokenType::Fun);
                }
            }
            break;
        case 'i':
            return checkKeyword(1, 1, "f", TokenType::If);
        case 'n':
            return checkKeyword(1, 2, "il", TokenType::Nil);
        case 'o':
            return checkKeyword(1, 1, "r", TokenType::Or);
        case 'p':
            return checkKeyword(1, 4, "rint", TokenType::Print);
        case 'r':
            return checkKeyword(1, 5, "eturn", TokenType::Return);
        case 's':
            return checkKeyword(1, 4, "uper", TokenType::Super);
        case 't':
            if (current - start > 1) {
                switch (start[1]) {
                    case 'h':
                        return checkKeyword(2, 2, "is", TokenType::This);
                    case 'r':
                        return checkKeyword(2, 2, "ue", TokenType::True);
                }
            }
            break;
        case 'v':
            return checkKeyword(1, 2, "ar", TokenType::Var);
        case 'w':
            return checkKeyword(1, 4, "hile", TokenType::While);
    }
    return TokenType::Identifier;
}

TokenType Lexer::checkKeyword(int start, int length, const char* rest, TokenType type) const {
    if (this->current - this->start == start + length && std::memcmp(this->start + start, rest, length) == 0) {
        return type;
    }
    return TokenType::Identifier;
}
bool Lexer::isEmpty() const {
    return source.empty();
}
void Lexer::setSource(const std::string& source) {

    this->source = source;
    start = source.c_str();
    current = start;
}

File: rclox/src/lexer.rs
use crate::error::RcloxError;
use crate::Result;

/// The type (or "kind", depending on who you ask) of a token
#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum TokenType {
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,

    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,

    Identifier,
    String,
    Number,

    And,
    Class,
    Else,
    False,
    For,
    Fun,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,

    Eof,
    Null,
}

/// A token of input. This has a type, a reference to the string it represents,
/// and a line number
#[derive(Debug, PartialEq, Eq)]
pub struct Token<'source_code> {
    pub kind: TokenType,
    pub span: &'source_code str,
    pub line: usize,
}

impl Token<'_> {
    /// An "eof" token
    pub const EOF: Self = Self {
        kind: TokenType::Eof,
        span: "",
        line: 0,
    };

    /// A "null" token
    pub const NULL: Self = Self {
        kind: TokenType::Null,
        span: "",
        line: 0,
    };
}

/// The lexer state machine
#[derive(Debug)]
pub struct Lexer<'source_code> {
    source: &'source_code [u8],
    start: usize,
    current: usize,
    line: usize,
}

impl<'source_code> Lexer<'source_code> {
    /// Construct a new, empty, lexer
    pub fn new(source: &'source_code str) -> Self {
        Lexer {
            source: source.as_bytes(),
            start: 0,
            current: 0,
            line: 0,
        }
    }

    /// Lex a token from the input, advancing the lexer appropriately
    pub fn lex_token(&'_ mut self) -> Result<Token<'source_code>> {
        self.skip_whitespace();
        self.start = self.current;

        if self.is_at_end() {
            return Ok(Token::EOF);
        };

        let c = self.advance();
        match c {
            '(' => Ok(self.make_token(TokenType::LeftParen)),
            ')' => Ok(self.make_token(TokenType::RightParen)),
            '{' => Ok(self.make_token(TokenType::LeftBrace)),
            '}' => Ok(self.make_token(TokenType::RightBrace)),
            ';' => Ok(self.make_token(TokenType::Semicolon)),
            ',' => Ok(self.make_token(TokenType::Comma)),
            '.' => Ok(self.make_token(TokenType::Dot)),
            '-' => Ok(self.make_token(TokenType::Minus)),
            '+' => Ok(self.make_token(TokenType::Plus)),
            '/' => Ok(self.make_token(TokenType::Slash)),
            '*' => Ok(self.make_token(TokenType::Star)),
            '!' => {
                if self.advance_if_char_matches('=') {
                    Ok(self.make_token(TokenType::BangEqual))
                } else {
                    Ok(self.make_token(TokenType::Bang))
                }
            }
            '=' => {
                if self.advance_if_char_matches('=') {
                    Ok(self.make_token(TokenType::EqualEqual))
                } else {
                    Ok(self.make_token(TokenType::Equal))
                }
            }
            '<' => {
                if self.advance_if_char_matches('=') {
                    Ok(self.make_token(TokenType::LessEqual))
                } else {
                    Ok(self.make_token(TokenType::Less))
                }
            }
            '>' => {
                if self.advance_if_char_matches('=') {
                    Ok(self.make_token(TokenType::GreaterEqual))
                } else {
                    Ok(self.make_token(TokenType::Greater))
                }
            }
            '"' => {
                while !self.is_at_end() && self.peek() != '"' {
                    if self.peek() == '\n' {
                        self.line += 1;
                    }
                    self.advance();
                }
                if self.is_at_end() {
                    return Err(RcloxError::Lexer("Unterminated string.".to_string()));
                }
                self.advance();
                Ok(self.make_token(TokenType::String))
            }
            '0'..='9' => {
                while self.peek().is_ascii_digit() {
                    self.advance();
                }
                if self.peek() == '.' && self.peek_next().map(|x| x.is_ascii_digit()).unwrap_or(false) {
                    self.advance();
                    while self.peek().is_ascii_digit() {
                        self.advance();
                    }
                }
                Ok(self.make_token(TokenType::Number))
            }
            'a'..='z' | 'A'..='Z' | '_' => {
                while self.peek().is_alphanumeric() || self.peek() == '_' {
                    self.advance();
                }
                let mut tok = self.make_token(TokenType::Null);
                tok.kind = match tok.span {
                    "and" => TokenType::And,
                    "class" => TokenType::Class,
                    "else" => TokenType::Else,
                    "false" => TokenType::False,
                    "for" => TokenType::For,
                    "fun" => TokenType::Fun,
                    "if" => TokenType::If,
                    "nil" => TokenType::Nil,
                    "or" => TokenType::Or,
                    "print" => TokenType::Print,
                    "return" => TokenType::Return,
                    "super" => TokenType::Super,
                    "this" => TokenType::This,
                    "true" => TokenType::True,
                    "var" => TokenType::Var,
                    "while" => TokenType::While,

                    _ => TokenType::Identifier,
                };
                Ok(tok)
            }

            _ => Err(RcloxError::Lexer(format!("Unexpected Character '{}'", c))),
        }
    }

    /// Check whether the lexer has consumed all of its input
    #[inline]
    fn is_at_end(&self) -> bool {
        self.current == self.source.len()
    }

    /// Advance the lexer over whitespace characters
    fn skip_whitespace(&mut self) {
        loop {
            match self.peek() {
                ' ' | '\r' | '\t' => {
                    self.advance();
                }
                '\n' => {
                    self.line += 1;
                    self.advance();
                }
                '/' => {
                    if self.peek_next() == Some('/') {
                        while self.peek() != '\n' && !self.is_at_end() {
                            self.advance();
                        }
                    } else {
                        return;
                    }
                }

                _ => return,
            };
        }
    }

    /// Advance the lexer if the current character matches a given character
    fn advance_if_char_matches(&mut self, expected: char) -> bool {
        if self.is_at_end() {
            return false;
        }
        if self.peek() != expected {
            return false;
        }
        self.current += 1;
        true
    }

    /// Peek the next character of the input
    fn peek_next(&self) -> Option<char> {
        self.source.get(self.current + 1).map(|&x| x.into())
    }

    /// Peek the current character of the input
    fn peek(&self) -> char {
        if self.current == self.source.len() {
            return 0.into();
        }
        self.source[self.current].into()
    }

    /// Advance the lexer, and return the next character
    fn advance(&mut self) -> char {
        self.current += 1;
        self.source[self.current - 1].into()
    }

    /// Construct a token with the given type with a span and a line
    fn make_token(&'_ self, kind: TokenType) -> Token<'source_code> {
        Token {
            kind,
            span: std::str::from_utf8(&self.source[self.start..self.current])
                .expect("Invalid byte slice encontered while creating token"),
            line: self.line,
        }
    }
}
